File: common/src/main/java/net/shiroha233/roadweaver/persistence/sqlite/RoadDatabaseManager.java
package net.shiroha233.roadweaver.persistence.sqlite;

import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.storage.LevelResource;

import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.Driver;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

/**
 * 旧 H2 读取入口，只供显式导入流程使用。
 */
final class RoadDatabaseManager {
    private RoadDatabaseManager() {}

    static final String DB_ROAD = "roads";
    static final String DB_TERRAIN = "terrain";
    static final String DB_MAP = "map";

    private static final String DB_DIR = "data/roadweaver";
    private static volatile boolean driverRegistered;

    static Connection openReadOnlyConnection(ServerLevel level, String dbName) throws SQLException {
        try {
            ensureDriverRegistered();
            Path dbPath = getDbPath(level, dbName);
            Files.createDirectories(dbPath.getParent());
            String url = "jdbc:h2:" + dbPath.toAbsolutePath()
                    + ";MODE=LEGACY"
                    + ";FILE_LOCK=NO"
                    + ";AUTO_SERVER=FALSE"
                    + ";CACHE_SIZE=8192"
                    + ";LOCK_MODE=0"
                    + ";ACCESS_MODE_DATA=r";

            Connection conn = DriverManager.getConnection(url, "sa", "");
            try (Statement stmt = conn.createStatement()) {
                stmt.execute("SET READONLY TRUE");
            } catch (SQLException ignored) {
            }
            return conn;
        } catch (Exception e) {
            throw new SQLException("Failed to create legacy H2 connection: " + dbName, e);
        }
    }

    static boolean hasAnyLegacyDatabase(ServerLevel level) {
        return hasLegacyDatabase(level, DB_ROAD)
                || hasLegacyDatabase(level, DB_TERRAIN)
                || hasLegacyDatabase(level, DB_MAP);
    }

    static boolean hasLegacyDatabase(ServerLevel level, String dbName) {
        if (level == null || dbName == null || dbName.isBlank()) return false;
        Path dbPath = getDbPath(level, dbName);
        return Files.exists(dbFile(dbPath, ".mv.db")) || Files.exists(dbFile(dbPath, ".h2.db"));
    }

    static void requireDriverAvailable() throws SQLException {
        ensureDriverRegistered();
    }

    static String dimKey(ServerLevel level) {
        ResourceLocation rl = level.dimension().location();
        return rl.getNamespace() + "_" + rl.getPath();
    }

    private static void ensureDriverRegistered() throws SQLException {
        if (driverRegistered) return;
        synchronized (RoadDatabaseManager.class) {
            if (driverRegistered) return;
            SQLException failure = null;
            for (String className : ListHolder.H2_DRIVER_CLASSES) {
                try {
                    Class<?> type = Class.forName(className);
                    Driver driver = (Driver) type.getDeclaredConstructor().newInstance();
                    DriverManager.registerDriver(driver);
                    driverRegistered = true;
                    return;
                } catch (Throwable t) {
                    if (failure == null) {
                        failure = new SQLException("H2 driver is not available");
                    }
                    failure.addSuppressed(t);
                }
            }
            throw failure == null ? new SQLException("H2 driver is not available") : failure;
        }
    }

    private static Path dbFile(Path dbPath, String suffix) {
        return dbPath.resolveSibling(dbPath.getFileName() + suffix);
    }

    private static Path getDbPath(ServerLevel level, String dbName) {
        Path worldRoot = level.getServer().getWorldPath(LevelResource.ROOT);
        return worldRoot.resolve(DB_DIR).resolve(dimKey(level)).resolve(dbName);
    }

    private static final class ListHolder {
        private static final String[] H2_DRIVER_CLASSES = {
                "org.h2.Driver",
                "net.shiroha233.roadweaver.libs.h2.Driver"
        };
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/persistence/sqlite/RoadDatabaseManager.kt
package net.shiroha233.roadweaver.persistence.sqlite

import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.storage.LevelResource
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Path
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.sql.Statement
import java.util.concurrent.ConcurrentHashMap

/**
 * SQLite 数据库连接管理器
 */
object RoadDatabaseManager {
    private val LOGGER = LoggerFactory.getLogger("roadweaver")

    @Volatile
    private var SQLITE_DRIVER_LOADED = false

    private val CONNECTIONS: ConcurrentHashMap<String, Connection> = ConcurrentHashMap()

    private const val DB_DIR = "data/roadweaver"
    private const val DB_NAME = "roads.db"

    private fun dimKey(level: ServerLevel): String {
        val rl: ResourceLocation = level.dimension().location()
        return rl.namespace + "_" + rl.path
    }

    private fun worldKey(level: ServerLevel): String {
        val worldRoot: Path? = level.server.getWorldPath(LevelResource.ROOT)
        val worldId = worldRoot?.toAbsolutePath()?.normalize()?.toString() ?: "unknown"
        return worldId + "|" + dimKey(level)
    }

    private fun getDbPath(level: ServerLevel): Path {
        val worldRoot = level.server.getWorldPath(LevelResource.ROOT)
        return worldRoot.resolve(DB_DIR).resolve(dimKey(level)).resolve(DB_NAME)
    }

    @Throws(SQLException::class)
    private fun ensureSqliteDriverLoaded() {
        if (SQLITE_DRIVER_LOADED) return
        synchronized(this) {
            if (SQLITE_DRIVER_LOADED) return
            try {
                Class.forName("org.sqlite.JDBC")
                SQLITE_DRIVER_LOADED = true
            } catch (e: ClassNotFoundException) {
                throw SQLException(
                    "SQLite JDBC driver not found. Dependency org.xerial:sqlite-jdbc may be missing.",
                    e
                )
            }
        }
    }

    /**
     * 获取或创建数据库连接
     */
    @JvmStatic
    @Throws(SQLException::class)
    fun getConnection(level: ServerLevel): Connection {
        val key = worldKey(level)

        var conn = CONNECTIONS[key]
        if (conn != null && !conn.isClosed) {
            return conn
        }

        synchronized(CONNECTIONS) {
            conn = CONNECTIONS[key]
            if (conn != null && !conn.isClosed) {
                return conn!!
            }

            try {
                val dbPath = getDbPath(level)
                Files.createDirectories(dbPath.parent)

                ensureSqliteDriverLoaded()

                val url = "jdbc:sqlite:" + dbPath.toAbsolutePath()
                conn = DriverManager.getConnection(url)

                conn!!.createStatement().use { stmt ->
                    stmt.execute("PRAGMA journal_mode=WAL")
                    stmt.execute("PRAGMA synchronous=NORMAL")
                    stmt.execute("PRAGMA cache_size=-8000")
                    stmt.execute("PRAGMA temp_store=MEMORY")
                    stmt.execute("PRAGMA foreign_keys=ON")
                }

                initTables(conn!!)

                CONNECTIONS[key] = conn!!
                LOGGER.debug("RoadDatabaseManager: 已创建维度 {} 的数据库连接", dimKey(level))

                try {
                    val migrated = LegacyShardMigration.migrateIfNeeded(level)
                    if (migrated > 0) {
                        LOGGER.info("RoadDatabaseManager: 维度 {} 已迁移 {} 条旧道路数据", dimKey(level), migrated)
                    }
                } catch (e: Exception) {
                    LOGGER.warn("RoadDatabaseManager: 旧数据迁移失败，不影响正常使用", e)
                }

                return conn!!
            } catch (e: Exception) {
                LOGGER.error("RoadDatabaseManager: 创建数据库连接失败", e)
                throw SQLException("Failed to create database connection", e)
            }
        }
    }

    @Throws(SQLException::class)
    private fun initTables(conn: Connection) {
        conn.createStatement().use { stmt ->
            stmt.execute(
                """
                CREATE TABLE IF NOT EXISTS roads (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    fingerprint INTEGER NOT NULL UNIQUE,
                    width INTEGER NOT NULL,
                    road_type INTEGER NOT NULL,
                    min_x INTEGER NOT NULL,
                    min_z INTEGER NOT NULL,
                    max_x INTEGER NOT NULL,
                    max_z INTEGER NOT NULL,
                    data BLOB NOT NULL,
                    created_at INTEGER DEFAULT (strftime('%s', 'now'))
                )
                """.trimIndent()
            )

            stmt.execute(
                """
                CREATE INDEX IF NOT EXISTS idx_roads_spatial
                ON roads (min_x, max_x, min_z, max_z)
                """.trimIndent()
            )

            stmt.execute(
                """
                CREATE INDEX IF NOT EXISTS idx_roads_fingerprint
                ON roads (fingerprint)
                """.trimIndent()
            )
        }
    }

    @JvmStatic
    fun closeConnection(level: ServerLevel) {
        val key = worldKey(level)
        val conn = CONNECTIONS.remove(key)
        if (conn != null) {
            try {
                conn.close()
                LOGGER.debug("RoadDatabaseManager: 已关闭维度 {} 的数据库连接", dimKey(level))
            } catch (e: SQLException) {
                LOGGER.warn("RoadDatabaseManager: 关闭数据库连接失败", e)
            }
        }
    }

    @JvmStatic
    fun closeAll() {
        for ((key, value) in CONNECTIONS.entries) {
            try {
                value.close()
            } catch (e: SQLException) {
                LOGGER.warn("RoadDatabaseManager: 关闭数据库连接失败: {}", key, e)
            }
        }
        CONNECTIONS.clear()
        LOGGER.debug("RoadDatabaseManager: 所有数据库连接已关闭")
    }

    @JvmStatic
    fun checkpoint(level: ServerLevel) {
        val key = worldKey(level)
        val conn = CONNECTIONS[key]
        if (conn != null) {
            try {
                conn.createStatement().use { stmt ->
                    stmt.execute("PRAGMA wal_checkpoint(TRUNCATE)")
                }
                LOGGER.debug("RoadDatabaseManager: 维度 {} WAL 检查点完成", dimKey(level))
            } catch (e: SQLException) {
                LOGGER.warn("RoadDatabaseManager: WAL 检查点失败", e)
            }
        }
    }

    @JvmStatic
    fun checkpointAll() {
        for ((key, conn) in CONNECTIONS.entries) {
            try {
                conn.createStatement().use { stmt ->
                    stmt.execute("PRAGMA wal_checkpoint(TRUNCATE)")
                }
            } catch (e: SQLException) {
                LOGGER.warn("RoadDatabaseManager: WAL 检查点失败: {}", key, e)
            }
        }
    }
}
